Fix golint failures: pkg/controller/certificates #68026

zerodayz opened the pull request:
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
This PR fix golint issues for pkg/controller/certificates:

Errors from golint:
pkg/controller/certificates/certificate_controller.go:43:6: exported type CertificateController should have comment or be unexported
pkg/controller/certificates/certificate_controller.go:57:1: exported function NewCertificateController should have comment or be unexported
pkg/controller/certificates/certificate_controller.go:212:52: exported func IgnorableError returns unexported type certificates.ignorableError, which can be annoying to use
pkg/controller/certificates/certificate_controller_utils.go:28:1: exported function GetCertApprovalCondition should have comment or be unexported

Which issue(s) this PR fixes:

Signed-off-by: Robin Cernin <cerninr@gmail.com>
#68026

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

soltysh commented on pkg/controller/certificates/certificate_controller.go:
```diff
@@ -209,12 +211,14 @@ func (cc *CertificateController) syncFunc(key string) error {
 
-type ignorableError string
+// NewIgnorableError error
+type NewIgnorableError string
```
I'd prefer we don't expose types we don't want to be exposed just because golint complains. Once made public it's part of the API we'll need to support. So I'd prefer not doing this.

soltysh reviewed (changes requested):
/hold

awly commented on pkg/controller/certificates/certificate_controller.go:
```diff
@@ -209,12 +211,14 @@ func (cc *CertificateController) syncFunc(key string) error {
 // sync should be retried.
-func IgnorableError(s string, args ...interface{}) ignorableError {
-	return ignorableError(fmt.Sprintf(s, args...))
+func IgnorableError(s string, args ...interface{}) NewIgnorableError {
```
This should return a regular `error` instead

zerodayz replied:
Thank you for your review and comments, I have modified the ignorableError so it only returns error.

awly commented on pkg/controller/certificates/certificate_controller.go:
```diff
@@ -40,6 +40,7 @@ import (
 	"k8s.io/kubernetes/pkg/controller"
 )
 
+// CertificateController holds the structure of the certificate controller
```
Describe what it does instead of what it holds

awly commented:
/ok-to-test

zerodayz commented:
Thank you @awly  I am still trying to understand parts of kubernetes code, so please correct me if I am wrong.

```
// IgnorableError returns an error that we shouldn't handle (i.e. log) because
// it's spammy and usually user error. Instead we will log these errors at a
// higher log level. We still need to throw these errors to signal that the
// sync should be retried.
func IgnorableError(s string, args ...interface{}) ignorableError {
	return ignorableError(fmt.Sprintf(s, args...))
}
```

Which seems to be only called from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/certificates/approver/sarapprove.go#L112

And here we return an unexported variable and that's what golint doesn't like, but now my understanding is that, is exactly what we want from it. We only export the function so it can be called from other package but we can't call the ignorableError directly.

When you said it should return regular error instead, did you mean something like this:

```
func IgnorableError(s string, args ...interface{}) error {
	return fmt.Errorf(s, args...)
}
```

I tried changing it and it returned as Error type while when using ignorableError, it returned as ignorableError type:

`*errors.errorString`
VS
`.ignorableError`

return ignorableError(fmt.Sprintf(s, args...))

I am not so much familiar with the code base but wonder, if this could cause any problems if we change it to regular error.

Thank you for your time reviewing this and answer.

awly commented:
@zerodayz it should look like so:

```
func IgnorableError(s string, args ...interface{}) error {
	return ignorableError(fmt.Sprintf(s, args...))
}
```

Which will work since `ignorableError` implements the `error` interface.
That way we both keep the internal error implementation hidden from the caller and keep the package API clean.

> One more question about the controller, do you know if there is any documentation to replace controller with custom controller, for debugging purposes?

This controller implementation is baked-in at compile time.
What are you trying to debug exactly? The answer will depend on that.

zerodayz commented:
> @zerodayz it should look like so:
> 
> 
> 
> ```
> 
> func IgnorableError(s string, args ...interface{}) error {
> 
> 	return ignorableError(fmt.Sprintf(s, args...))
> 
> }
> 
> ```
> 
> 
> 
> Which will work since `ignorableError` implements the `error` interface.
> 
> That way we both keep the internal error implementation hidden from the caller and keep the package API clean.
> 
> 
> 
> > One more question about the controller, do you know if there is any documentation to replace controller with custom controller, for debugging purposes?
> 
> 
> 
> This controller implementation is baked-in at compile time.
> 
> What are you trying to debug exactly? The answer will depend on that.

I am trying to debug controller’s workflow, instead of reading thru the code, I want to run it. It will be easier to also understand what the original developer of the code meant. Some of the code just doesn’t give sense to me and I want to understand what exactly is going on, without having to ask everytime.

I am already using kind but was wondering if there is a better way just to compile cert controller and replace image in deployment?

I know that I can use kubebuilder or operator sdk for custom controller, but I am unsure about controllers that are already part of k8s.

Thank you for your help and your time with me. I will do the changes and run tests and see if there are any problems, will submit new updated commit.

awly commented:
If I remember correctly, this controller runs in the [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) process.
One option is to re-compile and re-deploy `kube-controller-manager` (which sounds like what you've been doing so far).

The other option is to pass `--controllers=-csrapproving-csrsigning` flag to it (to disable the certificates controller) and run a separate controller process that you create. That separate process will take some work to create initially, but might be easier to iterate with.

zerodayz commented:
Thank you @awly for your help,

> If I remember correctly, this controller runs in the [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) process.
> One option is to re-compile and re-deploy `kube-controller-manager` (which sounds like what you've been doing so far).

I have been so far only re-compiling and re-deploying entire K8s cluster (using kind) not just individual component. I was thinking maybe there is an easier way, just like what you mentioned. (re-compile and re-deploy only kube-controller-manager) and replace in the deployment. Would that be possible?

> The other option is to pass `--controllers=-csrapproving-csrsigning` flag to it (to disable the certificates controller) and run a separate controller process that you create. That separate process will take some work to create initially, but might be easier to iterate with.

I am able to do the second part of running the controller separately from the cluster, but still have to find automated way to disable controllers (if possible with kind) or anything else. Want to avoid manual steps as much as possible.

awly commented:
Sorry @zerodayz, I have not used kind and don't have any good recommendations on how do iterate more quickly.

awly commented:
Ah, I see you've addressed the PR comments already.
Could you rebase the PR branch please?

zerodayz committed:
Fix golint failures: pkg/controller/certificates #68026

What type of PR is this?
/kind cleanup

What this PR does / why we need it:
This PR fix golint issues for pkg/controller/certificates:

Errors from golint:
pkg/controller/certificates/certificate_controller.go:43:6: exported type CertificateController should have comment or be unexported
pkg/controller/certificates/certificate_controller.go:57:1: exported function NewCertificateController should have comment or be unexported
pkg/controller/certificates/certificate_controller.go:212:52: exported func IgnorableError returns unexported type certificates.ignorableError, which can be annoying to use
pkg/controller/certificates/certificate_controller_utils.go:28:1: exported function GetCertApprovalCondition should have comment or be unexported

Which issue(s) this PR fixes:

Signed-off-by: Robin Cernin cerninr@gmail.com

zerodayz commented:
> Ah, I see you've addressed the PR comments already.
> Could you rebase the PR branch please?

Finished and also fixed another introduced golint issue, i guess since package was in .golint_failures there was no linting and new commits in certificate controller package

awly commented:
/lgtm

neolit123 commented:
/remove-area apiserver cloudprovider code-generation conformance dependency e2e-test-framework ipvs kubeadm kubectl kubelet provider/gcp release-eng test
/remove-sig api-machinery apps architecture autoscaling cli cloud-provider cluster-lifecycle instrumentation network node release scheduling storage testing

neolit123 commented:
/release-note-none
